File: bulk_renamer.go
// bulk_renamer.go
package main

import (
	"flag"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"time"
)

const (
	reset  = "\033[0m"
	green  = "\033[92m"
	red    = "\033[91m"
	yellow = "\033[93m"
	blue   = "\033[94m"
)

func colorize(text, color string) string {
	return color + text + reset
}

func getFiles(root string, recursive bool, extensions []string, filterRegex string) ([]string, error) {
	var files []string
	err := filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if info.IsDir() {
			if !recursive && path != root {
				return filepath.SkipDir
			}
			return nil
		}
		if extensions != nil {
			found := false
			for _, ext := range extensions {
				if strings.HasSuffix(info.Name(), "."+ext) {
					found = true
					break
				}
			}
			if !found {
				return nil
			}
		}
		if filterRegex != "" {
			matched, err := regexp.MatchString(filterRegex, info.Name())
			if err != nil {
				return err
			}
			if !matched {
				return nil
			}
		}
		files = append(files, path)
		return nil
	})
	return files, err
}

func applyTemplate(template, name, ext string, num int, prefix, suffix, replaceFrom, replaceTo, caseType string) string {
	newName := name
	if replaceFrom != "" {
		re := regexp.MustCompile(replaceFrom)
		newName = re.ReplaceAllString(newName, replaceTo)
	}
	switch caseType {
	case "lower":
		newName = strings.ToLower(newName)
	case "upper":
		newName = strings.ToUpper(newName)
	case "title":
		newName = strings.Title(newName)
	}
	if prefix != "" {
		newName = prefix + newName
	}
	if suffix != "" {
		newName = newName + suffix
	}

	// Замена переменных в шаблоне
	result := template
	re := regexp.MustCompile(`\{([^{}]+)\}`)
	result = re.ReplaceAllStringFunc(result, func(match string) string {
		key := strings.Trim(match, "{}")
		if key == "name" {
			return newName
		}
		if key == "ext" {
			return ext
		}
		if key == "old" {
			return name
		}
		if key == "date" {
			return time.Now().Format("2006-01-02")
		}
		if strings.HasPrefix(key, "num") {
			// поддерживаем формат {num:03d}
			if strings.Contains(key, ":") {
				parts := strings.SplitN(key, ":", 2)
				fmtStr := parts[1]
				return fmt.Sprintf(fmtStr, num)
			}
			return strconv.Itoa(num)
		}
		return match
	})
	return result
}

func main() {
	var (
		template   string
		path       string
		recursive  bool
		extStr     string
		filter     string
		prefix     string
		suffix     string
		replaceFrom string
		replaceTo   string
		caseType   string
		startNum   int
		dryRun     bool
		verbose    bool
	)
	flag.StringVar(&template, "t", "", "Шаблон нового имени")
	flag.StringVar(&path, "p", ".", "Путь к папке")
	flag.BoolVar(&recursive, "r", false, "Рекурсивно")
	flag.StringVar(&extStr, "e", "", "Фильтр по расширению (через запятую)")
	flag.StringVar(&filter, "f", "", "Фильтр по имени (regex)")
	flag.StringVar(&prefix, "prefix", "", "Префикс")
	flag.StringVar(&suffix, "suffix", "", "Суффикс")
	flag.StringVar(&replaceFrom, "replace-from", "", "Заменить от")
	flag.StringVar(&replaceTo, "replace-to", "", "Заменить на")
	flag.StringVar(&caseType, "case", "", "Регистр: lower, upper, title")
	flag.IntVar(&startNum, "number", 0, "Начать нумерацию с этого числа")
	flag.BoolVar(&dryRun, "dry-run", false, "Симуляция")
	flag.BoolVar(&verbose, "v", false, "Подробно")
	flag.Parse()

	if template == "" {
		fmt.Println(colorize("Ошибка: укажите шаблон (-t)", red))
		os.Exit(1)
	}

	var extensions []string
	if extStr != "" {
		extensions = strings.Split(extStr, ",")
	}

	files, err := getFiles(path, recursive, extensions, filter)
	if err != nil {
		fmt.Println(colorize("Ошибка: "+err.Error(), red))
		os.Exit(1)
	}

	if len(files) == 0 {
		fmt.Println(colorize("Нет файлов.", yellow))
		return
	}

	numCounter := startNum
	if numCounter == 0 {
		numCounter = 1
	}
	renamed := 0

	for _, fullPath := range files {
		dir := filepath.Dir(fullPath)
		oldName := filepath.Base(fullPath)
		ext := filepath.Ext(oldName)
		nameWithoutExt := strings.TrimSuffix(oldName, ext)
		ext = strings.TrimPrefix(ext, ".")

		newName := applyTemplate(template, nameWithoutExt, ext, numCounter, prefix, suffix, replaceFrom, replaceTo, caseType)
		numCounter++

		// Если расширение не было вставлено, добавляем
		if !strings.Contains(template, "{ext}") && ext != "" {
			newName = newName + "." + ext
		}

		newPath := filepath.Join(dir, newName)

		if dryRun || verbose {
			fmt.Printf("%s %s -> %s\n", colorize("->", blue), oldName, newName)
		}

		if !dryRun {
			// Проверка конфликта
			if _, err := os.Stat(newPath); err == nil && newPath != fullPath {
				// Добавляем суффикс
				base := strings.TrimSuffix(newName, filepath.Ext(newName))
				ext2 := filepath.Ext(newName)
				counter := 1
				for {
					tryPath := filepath.Join(dir, fmt.Sprintf("%s_%d%s", base, counter, ext2))
					if _, err := os.Stat(tryPath); os.IsNotExist(err) {
						newPath = tryPath
						break
					}
					counter++
				}
				if verbose {
					fmt.Println(colorize("  Конфликт, новое имя: "+filepath.Base(newPath), yellow))
				}
			}

			err := os.Rename(fullPath, newPath)
			if err != nil {
				fmt.Println(colorize("Ошибка при переименовании "+oldName+": "+err.Error(), red))
			} else {
				renamed++
			}
		}
	}

	if dryRun {
		fmt.Println(colorize(fmt.Sprintf("Симуляция завершена. Будет переименовано %d файлов.", len(files)), green))
	} else {
		fmt.Println(colorize(fmt.Sprintf("Переименовано %d файлов.", renamed), green))
	}
}

File: bulk_renamer.cs
// bulk_renamer.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;

class BulkRenamer
{
    static string Colorize(string text, string color)
    {
        string col = color switch
        {
            "green" => "\x1b[92m",
            "red" => "\x1b[91m",
            "yellow" => "\x1b[93m",
            "blue" => "\x1b[94m",
            _ => "\x1b[0m"
        };
        return col + text + "\x1b[0m";
    }

    static List<string> GetFiles(string root, bool recursive, List<string> extensions, string filterRegex)
    {
        var files = new List<string>();
        var dir = new DirectoryInfo(root);
        if (!dir.Exists) return files;

        var options = recursive ? SearchOption.AllDirectories : SearchOption.TopDirectoryOnly;
        foreach (var file in dir.GetFiles("*", options))
        {
            if (extensions != null && extensions.Any())
            {
                if (!extensions.Contains(file.Extension.TrimStart('.'), StringComparer.OrdinalIgnoreCase))
                    continue;
            }
            if (!string.IsNullOrEmpty(filterRegex))
            {
                if (!Regex.IsMatch(file.Name, filterRegex))
                    continue;
            }
            files.Add(file.FullName);
        }
        return files;
    }

    static string ApplyTemplate(string templateStr, string name, string ext, int num,
                                string prefix, string suffix, string replaceFrom, string replaceTo, string caseType)
    {
        string newName = name;
        if (!string.IsNullOrEmpty(replaceFrom))
        {
            newName = Regex.Replace(newName, replaceFrom, replaceTo);
        }
        switch (caseType?.ToLower())
        {
            case "lower": newName = newName.ToLower(); break;
            case "upper": newName = newName.ToUpper(); break;
            case "title": newName = System.Globalization.CultureInfo.CurrentCulture.TextInfo.ToTitleCase(newName); break;
        }
        if (!string.IsNullOrEmpty(prefix)) newName = prefix + newName;
        if (!string.IsNullOrEmpty(suffix)) newName = newName + suffix;

        // Replace variables
        string result = templateStr;
        var matches = Regex.Matches(result, @"\{([^{}]+)\}");
        foreach (Match m in matches)
        {
            string key = m.Groups[1].Value;
            string replacement = "";
            if (key == "name") replacement = newName;
            else if (key == "ext") replacement = ext;
            else if (key == "old") replacement = name;
            else if (key == "date") replacement = DateTime.Now.ToString("yyyy-MM-dd");
            else if (key.StartsWith("num"))
            {
                if (key.Contains(":"))
                {
                    string fmt = key.Split(':')[1];
                    replacement = num.ToString(fmt);
                }
                else
                {
                    replacement = num.ToString();
                }
            }
            else
            {
                replacement = m.Value;
            }
            result = result.Replace(m.Value, replacement);
        }
        return result;
    }

    static void Main(string[] args)
    {
        var opts = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        bool recursive = false, dryRun = false, verbose = false;
        int startNum = 1;

        for (int i = 0; i < args.Length; i++)
        {
            string arg = args[i];
            if (arg == "-t" && i+1 < args.Length) opts["template"] = args[++i];
            else if (arg == "-p" && i+1 < args.Length) opts["path"] = args[++i];
            else if (arg == "-r") recursive = true;
            else if (arg == "-e" && i+1 < args.Length) opts["ext"] = args[++i];
            else if (arg == "-f" && i+1 < args.Length) opts["filter"] = args[++i];
            else if (arg == "--prefix" && i+1 < args.Length) opts["prefix"] = args[++i];
            else if (arg == "--suffix" && i+1 < args.Length) opts["suffix"] = args[++i];
            else if (arg == "--replace-from" && i+1 < args.Length) opts["replace-from"] = args[++i];
            else if (arg == "--replace-to" && i+1 < args.Length) opts["replace-to"] = args[++i];
            else if (arg == "--case" && i+1 < args.Length) opts["case"] = args[++i];
            else if (arg == "--number" && i+1 < args.Length) startNum = int.Parse(args[++i]);
            else if (arg == "--dry-run") dryRun = true;
            else if (arg == "-v") verbose = true;
            else if (arg == "--help") { Console.WriteLine("Help..."); return; }
        }

        string template = opts.GetValueOrDefault("template", "");
        if (string.IsNullOrEmpty(template))
        {
            Console.WriteLine(Colorize("Error: template required (-t)", "red"));
            return;
        }

        string root = opts.GetValueOrDefault("path", ".");
        string extStr = opts.GetValueOrDefault("ext", "");
        string filter = opts.GetValueOrDefault("filter", "");
        string prefix = opts.GetValueOrDefault("prefix", "");
        string suffix = opts.GetValueOrDefault("suffix", "");
        string replaceFrom = opts.GetValueOrDefault("replace-from", "");
        string replaceTo = opts.GetValueOrDefault("replace-to", "");
        string caseType = opts.GetValueOrDefault("case", "");

        List<string> extensions = null;
        if (!string.IsNullOrEmpty(extStr))
            extensions = extStr.Split(',').Select(s => s.Trim()).ToList();

        var files = GetFiles(root, recursive, extensions, filter);
        if (files.Count == 0)
        {
            Console.WriteLine(Colorize("No files found.", "yellow"));
            return;
        }

        int numCounter = startNum;
        int renamed = 0;

        foreach (var fullPath in files)
        {
            string dir = Path.GetDirectoryName(fullPath);
            string oldName = Path.GetFileName(fullPath);
            string ext = Path.GetExtension(oldName).TrimStart('.');
            string nameWithoutExt = Path.GetFileNameWithoutExtension(oldName);

            string newName = ApplyTemplate(template, nameWithoutExt, ext, numCounter,
                                           prefix, suffix, replaceFrom, replaceTo, caseType);
            numCounter++;

            if (!template.Contains("{ext}") && !string.IsNullOrEmpty(ext))
            {
                newName = newName + "." + ext;
            }

            string newPath = Path.Combine(dir, newName);

            if (dryRun || verbose)
            {
                Console.WriteLine($"{Colorize("->", "blue")} {oldName} -> {newName}");
            }

            if (!dryRun)
            {
                // Conflict resolution
                if (File.Exists(newPath) && newPath != fullPath)
                {
                    string baseName = Path.GetFileNameWithoutExtension(newName);
                    string ext2 = Path.GetExtension(newName);
                    int counter = 1;
                    string tryPath;
                    do
                    {
                        tryPath = Path.Combine(dir, $"{baseName}_{counter}{ext2}");
                        counter++;
                    } while (File.Exists(tryPath));
                    newPath = tryPath;
                    if (verbose)
                    {
                        Console.WriteLine(Colorize($"  Conflict, new name: {Path.GetFileName(newPath)}", "yellow"));
                    }
                }

                try
                {
                    File.Move(fullPath, newPath);
                    renamed++;
                }
                catch (Exception e)
                {
                    Console.WriteLine(Colorize($"Error renaming {oldName}: {e.Message}", "red"));
                }
            }
        }

        if (dryRun)
            Console.WriteLine(Colorize($"Dry run completed. Would rename {files.Count} files.", "green"));
        else
            Console.WriteLine(Colorize($"Renamed {renamed} files.", "green"));
    }
}
